### PR TITLE
Util Response to #34

### PR DESCRIPTION
**Pull request by darioizzo**

### CODE REVIEW

File: doc/sphinx/docs/python/algorithms/py_algorithms.rst [hunk 1]
```diff
@@ -3,78 +3,136 @@
...
+PRAXIS (from nlopt)                                    :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+Nelder-Mead simplex (from nlopt)                       :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+sbplx (from nlopt)                                     :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+MMA (Method of Moving Asymptotes) (from nlopt)         :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+CCSA (from nlopt)                                      :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+SLSQP (from nlopt)                                     :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+low-storage BFGS (from nlopt)                          :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+preconditioned truncated Newton (from nlopt)           :class:`pygmo.nlopt`                      S-CU            Exposed from C++
+shifted limited-memory variable-metric (from nlopt)    :class:`pygmo.nlopt`                      S-CU            Exposed from C++
```
Comment (bluescarni) [inline doc/sphinx/docs/python/algorithms/py_algorithms.rst]: Missing the augmented Lagrangian algorithm. Also, BOBYQA, NEWUOA, PRAXIS, Nelder-Mead, sbplx do not handle constraints (only bound constraints). Same for BFGS, Newton and variable-metric.

File: doc/sphinx/docs/python/tutorials/evolving_a_population.rst [hunk 2]
```diff
@@ -12,8 +12,8 @@ an interior point optimizer or an evolutionary startegy solver, in pygmo you wil
 always have to call a method called ``evolve()`` to improve over your initial solutions,
 i.e. your *population*.
 
-The simplest way to evolve a population is to use directly the :class:`~pygmo.core.algorithm`
-method :class:`~pygmo.core.algorithm.evolve`
+The simplest way to evolve a population is to use directly the :class:`~pygmo.algorithm`
+method :func:`~pygmo.algorithm.evolve`
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/evolving_a_population.rst]: Usually we put trailing ``()`` when referring to functions (so there's immediate visual feedback that we are talking about a function): ``algorithm.evolve()``

File: doc/sphinx/docs/python/tutorials/evolving_a_population.rst [hunk 3]
```diff
@@ -41,12 +41,12 @@ method :class:`~pygmo.core.algorithm.evolve`
...
     :align: right
 
-Clearly, as :class:`~pygmo.core.sade` is a stochastic optimization algorithm, should we repeat the
+Clearly, as :class:`~pygmo.sade` is a stochastic optimization algorithm, should we repeat the
 evolution starting from the same population, we would obtain different results. If we
 want to monitor multiple runs from different initial populations and see how the final best fitness
 is achieved as the number of fitness evaluations increase. Most pygmo UDAs allow to do this
 as they maintain an internal log that can be accessed after the UDA has been
-extracted (see :class:`~pygmo.core.algorithm.extract`). This allows, for example to obtain
+extracted (see :class:`~pygmo.algorithm.extract`). This allows, for example to obtain
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/evolving_a_population.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 4]
```diff
@@ -99,15 +99,15 @@ point is (1.5, 0.75) etc. The plot below shows you the overall geometry of the e
 
 Once the hypervolume object is created, it allows for the computation of the following figures:
 
-1. :class:`~pygmo.core.hypervolume.compute` - Returns the joint hypervolume of the set of points (S-Metric).
+1. :func:`~pygmo.hypervolume.compute` - Returns the joint hypervolume of the set of points (S-Metric).
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 5]
```diff
@@ -99,15 +99,15 @@ point is (1.5, 0.75) etc. The plot below shows you the overall geometry of the e
...
 
 .. doctest::
 
     >>> # hv and ref_point refer to the data above
     >>> hv.compute(ref_point)
     3.25
 
-2. :class:`~pygmo.core.hypervolume.exclusive` - Returns the exclusive hypervolume by the point at given index.
+2. :func:`~pygmo.hypervolume.exclusive` - Returns the exclusive hypervolume by the point at given index.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 6]
```diff
@@ -119,15 +119,15 @@ Once the hypervolume object is created, it allows for the computation of the fol
     >>> hv.exclusive(3, ref_point)
     0.0
 
-3. :class:`~pygmo.core.hypervolume.least_contributor` - Returns the index of a point contributing the least to the hypervolume.
+3. :func:`~pygmo.hypervolume.least_contributor` - Returns the index of a point contributing the least to the hypervolume.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 7]
```diff
@@ -119,15 +119,15 @@ Once the hypervolume object is created, it allows for the computation of the fol
...
 
 .. doctest::
 
     >>> # hv and ref_point refer to the data above
     >>> hv.least_contributor(ref_point) # doctest: +SKIP
     3
 
-4. :class:`~pygmo.core.hypervolume.greatest_contributor` - Returns the index of a point contributing the most to the hypervolume.
+4. :func:`~pygmo.hypervolume.greatest_contributor` - Returns the index of a point contributing the most to the hypervolume.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 8]
```diff
@@ -138,11 +138,11 @@ Once the hypervolume object is created, it allows for the computation of the fol
 .. note::
   In case of several least/greatest contributors, pygmo returns only one contributor out of all candidates arbitrarily.
 
-5. :class:`~pygmo.core.hypervolume.contributions` - Returns a list of contributions for all points in the set.
-   This returns the same results as the successive call to the :class:`~pygmo.core.hypervolume.exclusive` method 
-   for each of the points. Due to the implementation, calling :class:`~pygmo.core.hypervolume.contributions` once can
+5. :func:`~pygmo.hypervolume.contributions` - Returns a list of contributions for all points in the set.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 9]
```diff
@@ -138,11 +138,11 @@ Once the hypervolume object is created, it allows for the computation of the fol
...
+   This returns the same results as the successive call to the :func:`~pygmo.hypervolume.exclusive` method
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/hypervolume.rst [hunk 10]
```diff
@@ -151,10 +151,10 @@ Once the hypervolume object is created, it allows for the computation of the fol
   array([ 0.5 ,  0.25,  0.5 ,  0.  ])
 
 Since all of the methods above require a reference point, it is useful to generate
-one automatically using the :class:`~pygmo.core.nadir`
+one automatically using the :func:`~pygmo.nadir`
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/hypervolume.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/moo.rst [hunk 11]
```diff
@@ -17,10 +17,10 @@ Let us start to define our population:
     >>> udp = zdt(prob_id = 1)
     >>> pop = population(prob = udp, size = 10, seed = 3453412)
 
-We here make use of first problem of the ZDT benchmark suite implemented in :class:`~pygmo.core.zdt` 
-and we create a :class:`~pygmo.core.population`
+We here make use of first problem of the ZDT benchmark suite implemented in :class:`~pygmo.zdt` 
+and we create a :class:`~pygmo.population`
 containing 10 individuals randomly created within the box bounds. Which individuals belong to which non-dominated front? 
-We can immediately see this by running the fast non-dominated sorting algorithm :func:`~pygmo.core.fast_non_dominated_sorting`:
+We can immediately see this by running the fast non-dominated sorting algorithm :func:`~pygmo.fast_non_dominated_sorting`:
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/moo.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/moo.rst [hunk 12]
```diff
@@ -57,9 +57,9 @@ where each successive pareto front is plotted in darker colour. If we now type:
     >>> ax = plot_non_dominated_fronts(pop.get_f()) # doctest: +SKIP
     >>> plt.title("ZDT1: ... and the evolved population")  # doctest: +SKIP
 
-we have instantiated the algorithm :class:`~pygmo.core.moead`, able to tackle
+we have instantiated the algorithm :class:`~pygmo.moead`, able to tackle
 multi-objective problems, fixing the number of generations to 250. In the following line we use directly
-the method :func:`~pygmo.core.moead.evolve` of the algorithm to evolve the :class:`~pygmo.core.population`
+the method :func:`~pygmo.moead.evolve` of the algorithm to evolve the :class:`~pygmo.population`
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/moo.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/moo_moead.rst [hunk 13]
```diff
@@ -1,16 +1,16 @@
...
+In this tutorial we will see how to use the user-defined algorithm :class:`~pygmo.moead` 
 provided by pygmo. In particular we will take as test cases problems from the DTLZ suite implemented
-in pygmo as the user-defined problem :class:`~pygmo.core.dtlz`.
+in pygmo as the user-defined problem :class:`~pygmo.dtlz`.
 
-The first, quick idea would be to instantiate the problem (say DTLZ1) and run :class:`~pygmo.core.moead`
+The first, quick idea would be to instantiate the problem (say DTLZ1) and run :class:`~pygmo.moead`
 with the default settings. We can monitor the convergence of the whole population to the Pareto front
-using the :func:`~pygmo.core.dtlz.p_distance` metric.
+using the :func:`~pygmo.dtlz.p_distance` metric.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/moo_moead.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/moo_moead.rst [hunk 14]
```diff
@@ -29,24 +29,24 @@ using the :func:`~pygmo.core.dtlz.p_distance` metric.
     >>> print(udp.p_distance(pop)) # doctest: +SKIP
     0.0012264939631066003
 
-Since the :func:`~pygmo.core.dtlz.p_distance` does not capture the information on the spread of the solutions we
-also compute the hypervolume indicator using the pygmo class :class:`~pygmo.core.hypervolume`:
+Since the :func:`~pygmo.dtlz.p_distance` does not capture the information on the spread of the solutions we
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/moo_moead.rst]: Ditto.

File: doc/sphinx/docs/python/tutorials/nlopt_basics.rst [hunk 15]
```diff
@@ -0,0 +1,124 @@
+.. _py_tutorial_nlopt_basics:
+
+A first tutorial on the use of NLOPT solvers
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/nlopt_basics.rst]: I capitalise ``NLopt`` throughout the code, following the style on their webpage.

File: doc/sphinx/docs/python/tutorials/nlopt_basics.rst [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+--------------------------------------------
+
+In this tutorial we show the basic usage pattern of :class:`pygmo.nlopt`. This user defined
+algorithm (UDA) wraps the NLOPT library making it easily accessible via the pygmo common
+:class:`pygmo.algorithm` interface. Let see how this miracle occur.
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/nlopt_basics.rst]: "Let's", "occurs"

File: doc/sphinx/docs/python/tutorials/nlopt_basics.rst [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+The lines above can be shortened and are equivalent to:
+
+.. doctest::
+   
+    >>> pop = pg.population(pg.luksan_vlcek1(dim = 20), size = 1)
+    >>> pop.problem.c_tol = [1e-8] * pop.problem.get_nc()
+
+.. image:: ../../images/nlopt_basic_lv1.png
+   :scale: 80 %
+   :alt: sade performance
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/nlopt_basics.rst]: Presumably "sade" -> "slsqp"?

File: doc/sphinx/docs/python/tutorials/using_problem.rst [hunk 18]
```diff
@@ -73,7 +73,7 @@ Lets check how a fitness computation increases the counter:
     True
 
 We may also get back the UDP, and thus access all the methods not exposed in the
-:class:`~pygmo.core.problem` interface at any time via the :class:`~pygmo.core.problem.extract` method:
+:class:`~pygmo.problem` interface at any time via the :class:`~pygmo.problem.extract` method:
```
Comment (bluescarni) [inline doc/sphinx/docs/python/tutorials/using_problem.rst]: Round brackets.

File: pygmo/__init__.py [hunk 19]
```diff
@@ -412,7 +412,8 @@ def _archi_init(self, n=0, **kwargs):
     """
     The constructor will initialise an archipelago with *n* islands built from *kwargs*.
     The keyword arguments accept the same format as explained in the constructor of
-    :class:`~pygmo.core.island`, with the following differences:
+
```
Comment (bluescarni) [inline pygmo/__init__.py]: Extra line slipping in?

File: pygmo/docstrings.cpp [hunk 20]
```diff
@@ -2782,11 +2781,260 @@ Runs the fast non dominated sorting algorithm on the input *points*
...
+
+)";
+}
+
+std::string non_dominated_front_2d_docstring()
+{
+    return R"(non_dominated_front_2d(points)
+
+Finds the non dominated front of a set of two dimensional objectives. Complexity is :math:`\mathcal{O}(N \log N)`
+and is thus lower than the complexity of calling :func:`~pygmo.fast_non_dominated_sorting`
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: Round brackets, missing period.

File: pygmo/docstrings.cpp [hunk 21]
```diff
@@ -2782,11 +2781,260 @@ Runs the fast non dominated sorting algorithm on the input *points*
...
+)";
+}
+
+std::string select_best_N_mo_docstring()
+{
+    return R"(select_best_N_mo(points, N)
+
+Returns (unordered) the best N individuals out of a multi-objective, unconstrained population, (intended here
+as a 2D array-like containing objective vectors). The strict ordering used is the same as that defined
+in :func:`~pygmo.sort_population_mo`
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: Ditto.

File: pygmo/docstrings.cpp [hunk 22]
```diff
@@ -2782,11 +2781,260 @@ Runs the fast non dominated sorting algorithm on the input *points*
...
+{
+    return R"(decompose_objectives(objs, weights, ref_point, method)
+
+Decomposes a vector of objectives.
+
+A vector of objectives is reduced to one only objective using a decomposition technique.
+
+Three different possibilities for *method* are here made available:
+
+- weighted decomposition,
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: Are these "-" interpreted correctly by sphinx as bullet lists?

File: pygmo/docstrings.cpp [hunk 23]
```diff
@@ -2782,11 +2781,260 @@ Runs the fast non dominated sorting algorithm on the input *points*
...
+z^*_n)` an :math:`n`-dimensional reference point. We also ussume :math:`\lambda_i > 0, \forall i=1..n` and :math:`\sum_i
+\lambda_i = 1`.
+
+The resulting single objective is thus defined as:
+
+- weighted decomposition: :math:`f_d(\mathbf x) = \boldsymbol \lambda \cdot \mathbf f`
+- Tchebycheff decomposition: :math:`f_d(\mathbf x) = \max_{1 \leq i \leq m} \lambda_i \vert f_i(\mathbf x) - z^*_i \vert`
+- boundary interception method (with penalty constraint): :math:`f_d(\mathbf x) = d_1 + \theta d_2`
+
+
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: Extra line?

File: pygmo/docstrings.cpp [hunk 24]
```diff
@@ -2782,11 +2781,260 @@ Runs the fast non dominated sorting algorithm on the input *points*
...
+Generates the requested number of weight vectors to be used to decompose a multi-objective problem. Three methods are available:
+
+- ``"grid"`` generates weights on an uniform grid. This method may only be used when the number of requested weights to be genrated is such that a uniform grid is indeed possible. 
+  In two dimensions this is always the case, but in larger dimensions uniform grids are possible only in special cases
+- ``"random"`` generates weights randomly distributing them uniformly on the simplex (weights are such that :math:`\sum_i \lambda_i = 1`) 
+- ``"low discrepancy"`` generates weights using a low-discrepancy sequence to, eventually, obtain a better coverage of the Pareto front. Halton sequence is used since
+  low dimensionalities are expected in the number of objectives (i.e. less than 20), hence Halton sequence is deemed as appropriate.
+
+.. note::  
+   All genration methods are guaranteed to generate weights on the simplex (:math:`\sum_i \lambda_i = 1`). All weight generation methods are guaranteed
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: "generation"

File: pygmo/docstrings.cpp [hunk 25]
```diff
@@ -2834,6 +3082,89 @@ Complexity is :math:`\mathcal{O}(MN)` where :math:`M` is the number of objective
...
+}
+
+std::string sort_population_con_docstring()
+{
+    return R"(sort_population_con(input_f, nec, tol)
+
+Sorts a population (intended here as a 2D array-like
+containing fitness vectors) assuming a single-objective, constrained case. 
+
+The following strict ordering is used (same as the one used in :func:`pygmo.compare_fc`):
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: Brackets.

File: pygmo/docstrings.cpp [hunk 26]
```diff
@@ -2879,7 +3211,7 @@ Performs a numerical estimation of the sparsity pattern of same callable object
 computing it around the input point *x* and detecting the components that are changed.
 
 The *callable* must accept an iterable as input and return an array-like object. The gradient returned will be dense
-and contain, in the lexicographic order requested by :class:`~pygmo.core.hypervoproblem.gradient`, :math:`\frac{df_i}{dx_j}`.
+and contain, in the lexicographic order requested by :class:`~pygmo.hypervoproblem.gradient`, :math:`\frac{df_i}{dx_j}`.
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: We have a class called hypervoproblem?

File: pygmo/docstrings.cpp [hunk 27]
```diff
@@ -3212,7 +3571,7 @@ std::string hv_refpoint_docstring()
 Calculates a mock refpoint by taking the maximum in each dimension over all points saved in the hypervolume object.
 The result is a point that is necessarily dominated by all other points, and thus can be used for hypervolume computations.
 
-**NOTE** This point is different from the one computed by :func:`~pygmo.core.nadir` as only the non dominated front is considered
+**NOTE** This point is different from the one computed by :func:`~pygmo.nadir` as only the non dominated front is considered
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: bracks

File: pygmo/docstrings.cpp (hunk 22 above)
Comment (darioizzo) [inline pygmo/docstrings.cpp]: yeps

**Pull request merged**